plutus-core/plutus-ir/src/PlutusIR/Transform/KnownCon.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+module PlutusIR.Transform.KnownCon (knownCon) where
+
+import PlutusCore qualified as PLC
+import PlutusCore.Name qualified as PLC
+import PlutusIR
+import PlutusIR.Core
+import PlutusIR.Transform.Rename ()
+
+import Control.Lens hiding (cons)
+import Data.Bitraversable
+import Data.List.Extra qualified as List
+import Data.Map (Map)
+import Data.Map qualified as Map
+
+{- | Simplify destructor applications, if the scrutinee is a constructor application.
+
+As an example, given
+
+@
+    Maybe_match
+      {x_type}
+      (Just {x_type} x)
+      {result_type}
+      (\a -> <just_case_body : result_type>)
+      (<nothing_case_body : result_type>)
+      additional_args
+@
+
+`knownCon` turns it into
+
+@
+    (\a -> <just_case_body>) x additional_args
+@
+-}
+knownCon ::
+    forall m tyname name uni fun a.
+    ( PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique
+    , Ord name
+    , PLC.MonadQuote m
+    ) =>
+    Term tyname name uni fun a ->
+    m (Term tyname name uni fun a)
+knownCon t0 = do
+    t <- PLC.rename t0
+    let ctxt =
+            foldMap
+                ( \case
+                    DatatypeBind _ (Datatype _ _ _ destr cons) ->
```

Comment:
If you wanted to be really clear, you could also record the number of type arguments, so you can be sure you're discarding the correct number when you discard them later. At the moment you're relying on noticing where it changes from type to term arguments, which I think _is_ fine, but counting the number of type arguments might be even more solid.

Reply:
Fair enough. Updated.